core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -74,134 +36,364 @@ class Queue[A] private (
+      def size: IO[E2, Int]                          = self.size
+      def shutdown: IO[E2, Unit]                     = self.shutdown
+      def take: IO[E2, C]                            = self.take.flatMap(f)
+      def takeAll: IO[E2, List[C]]                   = self.takeAll.flatMap(IO.foreach(_)(f))
+      def takeUpTo(max: Int): IO[E2, List[C]]        = self.takeUpTo(max).flatMap(IO.foreach(_)(f))
+    }
+
+  def zipWith[E2 >: E, A2 <: A, C, D](that: Queue2[E2, A2, C])(f: (B, C) => D): Queue2[E2, A2, D] =
+    new Queue2[E2, A2, D] {
+      def capacity: Int = math.min(self.capacity, that.capacity)
+
+      // TODO: atomicity? :-(
+      def offer(a: A2): IO[E2, Boolean]               = self.offer(a).zipWithPar(that.offer(a))(_ && _)
+      def offerAll(as: Iterable[A2]): IO[E2, Boolean] = self.offerAll(as).zipWithPar(that.offerAll(as))(_ && _)
+
+      def awaitShutdown: IO[E2, Unit] = self.awaitShutdown *> that.awaitShutdown
+      def size: IO[E2, Int]           = self.size.zipWithPar(that.size)(math.max)
+      def shutdown: IO[E2, Unit]      = self.shutdown.zipWithPar(that.shutdown)((_, _) => ())
+      def take: IO[E2, D]             = self.take.zipWith(that.take)(f)
+
+      def takeAll: IO[E2, List[D]] =
+        self.takeAll.zipWithPar(that.takeAll) { (bs, cs) =>
+          val bsIt = bs.iterator
+          val csIt = cs.iterator
+
+          bsIt.zip(csIt).map(f.tupled).toList
+        }
+
+      def takeUpTo(max: Int): IO[E2, List[D]] =
+        self.takeUpTo(max).zipWithPar(that.takeUpTo(max)) { (bs, cs) =>
+          val bsIt = bs.iterator
+          val csIt = cs.iterator
+
+          bsIt.zip(csIt).map(f.tupled).toList
+        }
+    }
 
-      as <- IO.sync0 { context =>
-             val as = unsafePollN(queue, max)
-             strategy.unsafeOnQueueEmptySpace(queue, context)
-             as
-           }
-    } yield as
+  def zip[E2 >: E, A2 <: A, C](that: Queue2[E2, A2, C]): Queue2[E2, A2, (B, C)] =
+    zipWith(that)((_, _))
+
+  def &&&[E2 >: E, A2 <: A, C](that: Queue2[E2, A2, C]): Queue2[E2, A2, (B, C)] =
+    zip(that)
+
+  def filter(f: B => Boolean): Queue2[E, A, B] =
+    new Queue2[E, A, B] {
+      def capacity: Int                             = self.capacity
+      def offer(a: A): IO[E, Boolean]               = self.offer(a)
+      def offerAll(as: Iterable[A]): IO[E, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E, Unit]                = self.awaitShutdown
+      // TODO: this doesn't account for filtered elements
+      def size: IO[E, Int]      = self.size
+      def shutdown: IO[E, Unit] = self.shutdown
+      def take: IO[E, B] = {
+        def retry: IO[E, B] =
+          self.take.flatMap { b =>
+            if (f(b)) IO.succeed(b)
+            else retry
+          }
+
+        retry
+      }
+
+      def takeAll: IO[E, List[B]] =
+        self.takeAll.map(_.filter(f))
+
+      def takeUpTo(max: Int): IO[E, List[B]] =
```

Comment:
@jdegoes I just noticed your comment in the issue re `filter` and building message hubs; this probably doesn't cut it because it is destructive.